unittests/algorithm.cpp

```diff
@@ -1213,17 +1219,117 @@ void parallel_transform2(size_t W) {
 }
 
 TEST_CASE("parallel_transform2.1thread") {
-  parallel_transform2(1);
+  parallel_transform2<std::vector<int>>(1);
+  parallel_transform2<std::list<int>>(1);
 }
 
 TEST_CASE("parallel_transform2.2threads") {
-  parallel_transform2(2);
+  parallel_transform2<std::vector<int>>(2);
+  parallel_transform2<std::list<int>>(2);
 }
 
 TEST_CASE("parallel_transform2.3threads") {
-  parallel_transform2(3);
+  parallel_transform2<std::vector<int>>(3);
+  parallel_transform2<std::list<int>>(3);
 }
 
 TEST_CASE("parallel_transform2.4threads") {
-  parallel_transform2(4);
+  parallel_transform2<std::vector<int>>(4);
+  parallel_transform2<std::list<int>>(4);
+}
+
+void parallel_transform3(size_t W) {
+
+  std::srand(static_cast<unsigned int>(time(NULL)));
+
+  tf::Taskflow taskflow;
+  tf::Executor executor(W);
+
+  using std::string;
+  using std::size_t;
+
+  for(size_t N=0; N<1000; N++) {
+
+    std::multimap<int, size_t> src;
+
+    /** Reference implementation with std::transform */
+    std::vector<string> ref;
+
+    /** Target implementation with Subflow::transform */
+    std::vector<string> tgt;
+
+    /** A generic function to cast integers to string */
+    const auto myFunction = [](const size_t x) -> string {
+      return "id_" + std::to_string(x);
+    };
+
+    taskflow.clear();
+
+    /** Group integers 0..(N-1) into ten groups,
+     * each having an unique key `d`.
+     */
+    auto from = taskflow.emplace([&, N](){
+      for(size_t i = 0; i < N; i++) {
+        const int d = ::rand() % 10;
+        src.emplace(d, i);
+      }
+    });
+
+    auto to_ref = taskflow.emplace([&, N]() {
+
+      // Find entries matching key = 0.
+      // This can return empty results.
+      const auto [src_beg, src_end] = src.equal_range(0);
+      const size_t n_matching = std::distance(src_beg, src_end);
+      ref.resize(n_matching);
+
+      // Extract all values having matching key value.
+      std::transform(src_beg, src_end, ref.begin(),
+        [&](const auto& x) -> string {
+          return myFunction(x.second);
+      });
+    });
+
+    /** Dynamic scheduling with Subflow::transform */
+    auto to_tgt = taskflow.emplace([&, N](tf::Subflow& subflow) {
+
+      // Find entries matching key = 0
+      const auto [src_beg, src_end] = src.equal_range(0);
+      const size_t n_matching = std::distance(src_beg, src_end);
+      tgt.resize(n_matching);
+
+      subflow.transform(std::ref(src_beg), std::ref(src_end), std::ref(tgt.begin()),
+        [&] (const auto& x) -> string {
+          return myFunction(x.second);
+      });
+    });
+
```

Comment:
Hi @tsung-wei-huang , I turned this PR from "draft" to "ready" as you instructed.

I am also half way through hammering out the test case for `Subflow::transform`. Currently the compiler complains about this error. I will debug my test code later this evening.

```c++
FAILED: unittests/CMakeFiles/algorithm.dir/algorithm.cpp.o 
/usr/bin/c++  -I/home/antony/Documents/taskflow -I/home/antony/Documents/taskflow/3rd-party/doctest -O3 -DNDEBUG -Wall -Wextra -Wfatal-errors -std=gnu++1z -MD -MT unittests/CMakeFiles/algorithm.dir/algorithm.cpp.o -MF unittests/CMakeFiles/algorithm.dir/algorithm.cpp.o.d -o unittests/CMakeFiles/algorithm.dir/algorithm.cpp.o -c /home/antony/Documents/taskflow/unittests/algorithm.cpp
/home/antony/Documents/taskflow/unittests/algorithm.cpp: In lambda function:
/home/antony/Documents/taskflow/unittests/algorithm.cpp:1301:87: error: use of deleted function ‘void std::ref(const _Tp&&) [with _Tp = __gnu_cxx::__normal_iterator<std::__cxx11::basic_string<char>*, std::vector<std::__cxx11::basic_string<char> > >]’
       subflow.transform(std::ref(src_beg), std::ref(src_end), std::ref(tgt.begin()),
                                                                                   ^
```

Reply:
Thanks. I don't think you can `std::ref` a r-value (`tgt.begin()`), which may be the reason of compiler error.

Reply:
@tsung-wei-huang I finished debugging the test code. Could you please review the code again? 

`cmake -GNinja . build/ && ninja -C build/ test` returns 100% test pass.